### PR TITLE
Upgrade Gradle to 7.1

### PR DESCRIPTION
**Pull request by ikhoon**
Motivation:

Gradle 7.1 has been released. https://docs.gradle.org/7.1/release-notes.html

Modifications:

- Add an explicit dependency for implicit dependency between sourcesJar
  and generated sources.
  Many warnings similar to the following messages produced while building
  projects. See https://docs.gradle.org/7.1/userguide/validation_problems.html#implicit_dependency
  ```
  Execution optimizations have been disabled for task ':grpc:generateProto' to ensure correctness due to the following reasons:
  - Gradle detected a problem with the following location: '/Users/ikhun/src/armeria/grpc/gen-src/main'. Reason: Task ':grpc:sourcesJar' uses this output of task ':grpc:generateProto' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.1/userguide/validation_problems.html#implicit_dependency for more details about this problem.
  - Gradle detected a problem with the following location: '/Users/ikhun/src/armeria/grpc/gen-src/main/resources/META-INF/armeria/grpc/armeria-main.dsc'. Reason: Task ':grpc:sourcesJar' uses this output of task ':grpc:generateProto' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.1/userguide/validation_problems.html#implicit_dependency for more details about this problem.
  ```
- Add `kotlin` flag and dedupe Kotlin configurations
- Remove `jcenter()` which has been deprecated
- Replace `mainClassName` witth `mainClass.set()`
- Upgrade groovy-xml from 2.5.12 -> 3.0.5
- Specify the default `DuplicatesStrategy` to `EXCLUDE`
  - The duplicated files raised error while build projects

Result:

Bump Gradle version from 6.8.3 to 7.1

### CODE REVIEW

**Comment by trustin** [comment]
> Specify the default `DuplicatesStrategy` to `EXCLUDE`

Wouldn't this conflict with our version properties files?

**Comment by ikhoon** [comment]
> > Specify the default `DuplicatesStrategy` to `EXCLUDE`
> 
> Wouldn't this conflict with our version properties files?

There are some resource files that conflict. version properties are also included. I thought `EXCLUDE` strategy(first win) seems to work for us.
But I was not sure it is the right approach for this problem.

**Comment by trustin** [comment]
Looks like we still have some warnings left to address:

- https://github.com/line/armeria/pull/3639/checks?check_run_id=2854952055#step:7:1439
- https://github.com/line/armeria/pull/3639/checks?check_run_id=2854952017#step:7:350
- https://github.com/line/armeria/pull/3639/checks?check_run_id=2854952017#step:7:1070
- https://github.com/line/armeria/pull/3639/checks?check_run_id=2854952072#step:7:60
- https://github.com/line/armeria/pull/3639/checks?check_run_id=2854952072#step:7:71
- https://github.com/line/armeria/pull/3639/checks?check_run_id=2854952072#step:7:225

But.. can we just disable this warning at all? Our project is not that large so this optimization warning doesn't really matter to us at the moment.

**Comment by ikhoon** [comment]
> But.. can we just disable this warning at all? Our project is not that large so this optimization warning doesn't really matter to us at the moment.

Unfortunately, it seems like there is no such option only for disabling this implicit dependency warning.

**Comment by ikhoon** [comment]
Opened a PR for gradle-scalafmt to clean up the warning messages. https://github.com/alenkacz/gradle-scalafmt/pull/43

File: gradle/scripts/lib/java-rpc-thrift.gradle [hunk 1]
```diff
@@ -21,6 +21,9 @@ configure(projectsWithFlags('java')) {
                 group: 'Build',
                 description: "Compiles the ${sourceSet.name} .thrift files."
         ])
+        if (scope == 'main') {
+            project.sourcesJar.dependsOn(task)
+        }
```
Comment (trustin) [inline gradle/scripts/lib/java-rpc-thrift.gradle:26]:
```suggestion
        def sourceJarTask = projects.tasks.findByName(sourceSet.getTaskName('sources', 'jar'))
        if (sourceJarTask) {
            sourceJarTask.dependsOn(task)
        }
```

Please make a similar change to `java-rpc-proto.gradle`.

File: gradle/scripts/lib/kotlin.gradle [hunk 2]
```diff
@@ -0,0 +1,35 @@
+configure(projectsWithFlags('kotlin')) {
```
Comment (trustin) [inline gradle/scripts/lib/kotlin.gradle:1]: :+1:

File: build.gradle [hunk 3]
```diff
@@ -162,11 +161,13 @@ configure(projectsWithFlags('java')) {
...
-            jvmTarget = "1.8"
-            freeCompilerArgs = ["-Xjsr305=strict"]
-        }
+    // Configure default duplicates strategy for such as:
+    // - c.l.armeria.versions.properties
+    // - thrift/cassandra.json
+    // - and so on
+    tasks.sourcesJar.duplicatesStrategy = DuplicatesStrategy.EXCLUDE
+    tasks.withType(Copy) {
+        duplicatesStrategy = DuplicatesStrategy.EXCLUDE
```
Comment (trustin) [inline build.gradle:170]: So, should we use `INCLUDE` here?

Comment (ikhoon) [reply build.gradle:170]: That makes sense. A file could be updated on purpose.

**Review by trustin — APPROVED** [review]

**Pull request merged**